src/defineFunction.js

```diff
@@ -1,94 +1,127 @@
-
-        // Whether or not the function is allowed inside text mode
-        // (default true)
-        allowedInMath?: boolean,
-
-        // (optional) The number of optional arguments the function
-        // should parse. If the optional arguments aren't found,
-        // `null` will be passed to the handler in their place.
-        // (default 0)
-        numOptionalArgs?: number,
-
-        // Must be true if the function is an infix operator.
-        infix?: boolean,
-    },
+    props: FunctionPropSpec,
 
     // The handler is called to handle these functions and their arguments.
-    // It receives two arguments:
-    //   - context contains information and references provided by the parser
-    //   - args is an array of arguments obtained from TeX input
-    // The context contains the following properties:
-    //   - funcName: the text (i.e. name) of the function, including \
-    //   - parser: the parser object
-    //   - lexer: the lexer object
-    // The latter three should only be used to produce error messages.
     //
     // The function should return an object with the following keys:
     //   - type: The type of element that this is. This is then used in
     //          buildHTML/buildMathML to determine which function
     //          should be called to build this node into a DOM node
     // Any other data can be added to the object, which will be passed
     // in to the function in buildHTML/buildMathML as `group.value`.
-    handler: (context: any, args: any) => T,
+    handler: ?FunctionHandler,
 
     // This function returns an object representing the DOM structure to be
     // created when rendering the defined LaTeX function.
-    htmlBuilder: (group: T, options: Options) => any,
+    // TODO: Port buildHTML to flow and make the group and return types explicit.
+    htmlBuilder?: (group: *, options: Options) => *,
 
     // This function returns an object representing the MathML structure to be
     // created when rendering the defined LaTeX function.
-    mathmlBuilder: (group: T, options: Options) => any,
-}
+    // TODO: Port buildMathML to flow and make the group and return types explicit.
+    mathmlBuilder?: (group: *, options: Options) => *,
+|};
+
+/**
+ * Final function spec for use at parse time.
+ * This is almost identical to `FunctionPropSpec`, except it
+ * 1. includes the function handler, and
+ * 2. requires all arguments except argTypes.
+ * It is generated by `_defineFunction`.
+ */
+type FunctionSpec = {|
+    numArgs: number,
+    argTypes?: ArgType[],
+    greediness: number,
+    allowedInText: boolean,
+    allowedInMath: boolean,
+    numOptionalArgs: number,
+    infix: boolean,
+    // Must be specified unless it's handled directly in the parser.
+    handler: ?FunctionHandler,
+|};
+
+/**
+ * All registered functions. Import this via 'functions.js' since that ensures
+ * that all functions are defined.
+ */
+export const _functions: {[string]: FunctionSpec} = {};
```

Comment:
This should be `functions` since it's a public.

Reply:
I just did this since we don't want _anyone_ but `functions.js` to import this. I updated the comment above this to emphasize that fact a bit more.
I could also name this to `importOnlyInFunctionsJs` or `functionsDoNotImportDirectly` or some such to make it clearer. ;)

Reply:
`functions.js` just exports this same dictionary again and makes it public.  `Parser.js` requires this dictionary.

Reply:
Done.

Reply:
The comment is good, but it doesn't prevent people from access the `functions` map.  In the past we haven't really had issues with people accessing `functions`, but if you're concerned that people might do so in the future then maybe we should take additional steps.  One possibility might be to have functions.js export a `getFunction` method instead of `functions` itself.

Reply:
1) I'm sorry, did I merge before this was resolved? If so, sorry!
2) I'd very much considered wrapping this in a method, although I was thinking `getFunctions` (which has no benefits over this) as opposed to `getFunction`, which would indeed be better. We could do this as follow-ups as we refactor further.

Reply:
It's my fault for hitting "accept" early.

Reply:
Ah, I'll just let you merge it next time.
Perhaps I have an unreasonable fear of merge conflicts. ;)